Best effort support for cross-architecture syscalls

Keno opened the pull request:
As mentioned on IRC, when using `int $0x80`, the kernel behaves as if the process were 32bit, contrary to rr's expectation, which is that it's just a different spelling of `syscall`. This adds some best-effort code to handle this. In my testing this works surprisingly well, and runs a couple of
the test cases I had tried, but I made no effort or attempt at rigorous testing of this functionality. There's a new warning message that gets printed in the log when a cross-arch syscall is detected though, so we should be able to figure out what's going on if we come across any real applications we care about that do this.

While I'm at it (it being weird corners of the x86 architecture), also clean up the modify_ldt support commit I had lying around locally. With both of these commits, I was also able to successfully record&replay the [ldt_gdt kernel selftest](https://github.com/torvalds/linux/blob/master/tools/testing/selftests/x86/ldt_gdt.c), which uses both of these features.

rocallahan commented on src/Task.cc:
```diff
@@ -1310,42 +1314,70 @@ static bool is_in_non_sigreturn_exit_syscall(Task* t) {
+  if (t->session().is_recording()) {
+    auto rt = static_cast<RecordTask*>(t);
+    if (rt->ev().is_syscall_event())
+      return rt->ev().Syscall().arch();
```
{}

rocallahan commented on src/Task.cc:
```diff
@@ -1310,42 +1314,70 @@ static bool is_in_non_sigreturn_exit_syscall(Task* t) {
+  } else if (t->session().is_replaying()) {
+    auto rt = static_cast<ReplayTask*>(t);
+    if (rt->current_trace_frame().event().is_syscall_event())
+      return rt->current_trace_frame().event().Syscall().arch();
```
{}

rocallahan commented on src/record_syscall.cc:
```diff
@@ -4818,14 +4820,20 @@ static void rec_process_syscall_arch(RecordTask* t,
-  rec_process_syscall_internal(t, syscall_state);
+  SupportedArch arch = t->ev().arch();
+  if (arch != t->arch())
+    LOG(warn) << "Cross architecture syscall detected. Support is best effort";
```
{}

rocallahan replied:
I'm not sure it's really a great idea to print these on every syscall. Maybe set a global flag so this only gets printed once per rr run or something?

rocallahan commented on src/record_syscall.cc:
```diff
@@ -3749,6 +3749,20 @@ static Switchable rec_prepare_syscall_arch(RecordTask* t,
+      int func = regs.arg1();
+      if (func == 0 || func == 2) {
+        syscall_state.reg_parameter(
+            2, ParamSize::from_syscall_result<typename Arch::ssize_t>(
```
Is `ssize_t` correct? The man page says `int`.

Keno replied:
Good catch.

Keno replied:
Upon reading the man page again, I also realized that unlike set_thread_area, modify_ldt does not poke `entry_number`, so I can the else branch here.

Keno commented:
> I assume you have an actual need for this?

"Need" is a strong word. I came across an application I wanted to debug that used `int $0x80` in 64bit mode and rr did the wrong thing. I'd rather rr had done the right thing (which it does now). Of course any applications using `int $0x80` in 64bit mode are probably doing something wrong, but as a debugger, we probably can't be too picky there.

rocallahan commented:
> I came across an application I wanted to debug that used int $0x80 in 64bit mode and rr did the wrong thing.

Fine, that's enough for our purposes.

In case you didn't see it --- your test failed on the buildbot.

Keno commented:
Yep, I have a handle on it. Fix coming (will have to wait for after dinner though - didn't plan on you reviewing so quickly ;) )

Keno commented:
Ok. I've updated this. Sorry this took a while. The builbot failure turned out a little more tricky than I had hoped, because we didn't have the requisite information about which syscall we were at anymore. Instead I took `emulate_syscall_entry` (which I just not realize I meant to rename to something more appropriate to what it does, since it's not only needed on entry - will push an update to that extent in a minute) and put it in the place where it's actually needed, rather than doing it in `did_waitpid`. That feels a lot less hacky to me and fixes the buildbot issue for free.

What had me going for a while was an intermittent failure due to a bug I introduced and then inadvertently fixed. Unfortunately because of how rare the failure was, I didn't realize I had inadvertently fixed it, so spent longer than I wish to admit trying to write a deterministic test case for a bug that was already fixed, oh well.

Keno commented:
Argh. Still failing on the buildbot. Different test though. Don't think I've seen that one. Will take a look.

Keno commented:
Ok, was able to reproduce this. It is very intermittent, but I think that test is just racey (in particular the thread can exit before we enter the futex wait). I think we should probably just change it to allow EAGAIN (unless somebody can think of a better solution).

rocallahan commented:
Great!
